src/coreclr/jit/morph.cpp

```diff
@@ -12387,6 +12402,72 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
 
 #endif
 
+//------------------------------------------------------------------------
+// fgOptimizeDivision: Optimizes division operations.
+//
+// Arguments:
+//   tree - the unchecked GT_DIV/GT_UDIV tree to optimize.
+//
+// Return Value:
+//   The optimized tree that can have any shape.
+//
+GenTree* Compiler::fgOptimizeDivision(GenTreeOp* div)
+{
+    assert(div->OperIs(GT_DIV, GT_UDIV));
+
+    if (gtTreeHasSideEffects(div, GTF_SIDE_EFFECT))
```

Comment:
you only need to take side effects into account for the case where you fold to 0, and i guess you need to check op1 there